// @nx/js/src/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createPackageJson = exports.getLockFileName = exports.createLockFile = exports.isValidVariable = exports.setupVerdaccio = exports.setupPrettierGenerator = exports.initGenerator = exports.libraryGenerator = void 0;
const tslib_1 = require("tslib");
tslib_1.__exportStar(require("./utils/typescript/add-tslib-dependencies"), exports);
tslib_1.__exportStar(require("./utils/typescript/load-ts-transformers"), exports);
tslib_1.__exportStar(require("./utils/typescript/print-diagnostics"), exports);
tslib_1.__exportStar(require("./utils/typescript/run-type-check"), exports);
tslib_1.__exportStar(require("./utils/typescript/get-source-nodes"), exports);
tslib_1.__exportStar(require("./utils/compiler-helper-dependency"), exports);
tslib_1.__exportStar(require("./utils/typescript/ts-config"), exports);
tslib_1.__exportStar(require("./utils/typescript/create-ts-config"), exports);
tslib_1.__exportStar(require("./utils/typescript/ast-utils"), exports);
tslib_1.__exportStar(require("./utils/package-json"), exports);
tslib_1.__exportStar(require("./utils/assets"), exports);
tslib_1.__exportStar(require("./utils/package-json/update-package-json"), exports);
tslib_1.__exportStar(require("./utils/package-json/create-entry-points"), exports);
var library_1 = require("./generators/library/library");
Object.defineProperty(exports, "libraryGenerator", { enumerable: true, get: function () { return library_1.libraryGenerator; } });
var init_1 = require("./generators/init/init");
Object.defineProperty(exports, "initGenerator", { enumerable: true, get: function () { return init_1.initGenerator; } });
var generator_1 = require("./generators/setup-prettier/generator");
Object.defineProperty(exports, "setupPrettierGenerator", { enumerable: true, get: function () { return generator_1.setupPrettierGenerator; } });
var generator_2 = require("./generators/setup-verdaccio/generator");
Object.defineProperty(exports, "setupVerdaccio", { enumerable: true, get: function () { return generator_2.setupVerdaccio; } });
var is_valid_variable_1 = require("./utils/is-valid-variable");
Object.defineProperty(exports, "isValidVariable", { enumerable: true, get: function () { return is_valid_variable_1.isValidVariable; } });
// eslint-disable-next-line @typescript-eslint/no-restricted-imports
var lock_file_1 = require("nx/src/plugins/js/lock-file/lock-file");
Object.defineProperty(exports, "createLockFile", { enumerable: true, get: function () { return lock_file_1.createLockFile; } });
Object.defineProperty(exports, "getLockFileName", { enumerable: true, get: function () { return lock_file_1.getLockFileName; } });
// eslint-disable-next-line @typescript-eslint/no-restricted-imports
var create_package_json_1 = require("nx/src/plugins/js/package-json/create-package-json");
Object.defineProperty(exports, "createPackageJson", { enumerable: true, get: function () { return create_package_json_1.createPackageJson; } });
